### PR TITLE
[Feature] Support pixel reconstruction visualization

### PR DESCRIPTION
**Pull request by fangyixiao18**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

To refactor the feature extraction pipeline and support pixel reconstruction visualization, such as MAE, SimMIM and MaskFeat
Related issue: https://github.com/open-mmlab/mmselfsup/issues/553

## Modification

Refactor Inference API
Add function `reconstruct` for MAE, SimMIM and MaskFeat algorithm python file

demo image for visualzation:
MAE:
test_mae
SimMIM:
test_simmim
MaskFeat:
test_maskfeat_RGB
test_maskfeat

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: docs/en/user_guides/analysis_tools.md [hunk 1]
```diff
@@ -142,3 +143,53 @@ An example of visualization:
 <div align="center">
 <img src="https://user-images.githubusercontent.com/36138628/199388251-476a5ad2-f9c1-4dfb-afe2-73cf41b5793b.jpg" width="800" />
 </div>
+
+## Pixel Reconstruction Visualization
```
Comment (YuanLiuuuuuu) [inline docs/en/user_guides/analysis_tools.md]:
```suggestion
## Low-level Feature Reconstruction Visualization
```

File: docs/en/user_guides/visualization.md [hunk 2]
```diff
@@ -9,6 +9,9 @@ Visualization can give an intuitive interpretation of the performance of the mod
   - [What Visualization do in MMSelfsup](#what-visualization-do-in-mmselfsup)
   - [Use different storage backends](#use-different-storage-backends)
```
Comment (tonysy) [inline docs/en/user_guides/visualization.md]:
```suggestion
  - [Use Different Storage Backends](#use-different-storage-backends)
```

File: docs/en/user_guides/visualization.md [hunk 3]
```diff
@@ -86,3 +89,103 @@ E.g.
...
+
+- `CONFIG_FILE`: config file for the pre-trained model.
+- `CKPT_PATH`: the path of model's checkpoint.
+- `WORK_DIR`: the directory to save the results of visualization.
+- `[optional arguments]`: for optional arguments, you can refer to [visualize_tsne.py](https://github.com/open-mmlab/mmselfsup/blob/master/tools/analysis_tools/visualize_tsne.py)
+
+An example:
+
+```shell
+python tools/analysis_tools/visualize_tsne.py configs/selfsup/simsiam/simsiam_resnet50_8xb32-coslr-100e_in1k.py --checkpoint epoch_100.pth --work-dir work_dirs/selfsup/simsiam_resnet50_8xb32-coslr-200e_in1k
```
Comment (tonysy) [inline docs/en/user_guides/visualization.md:134]: 是不是可以加个 wget link，把我们提供的模型的填上，让用户直接执行命令可以跑，这个epoch_100 得自己训练吧

Comment (tonysy) [reply docs/en/user_guides/visualization.md:134]: 或者是传http链接

File: docs/en/user_guides/visualization.md [hunk 4]
```diff
@@ -86,3 +89,103 @@ E.g.
...
+We provide several reconstruction visualization for listed algorithms:
+
+- MAE
+- SimMIM
+- MaskFeat
+
+Users can run command below to visualize the reconstruction.
+
+```shell
+python tools/analysis_tools/visualize_reconstruction.py ${CONFIG_FILE} --checkpoint ${CKPT_PATH} --img-path ${IMAGE_PATH} --out-file ${OUTPUT_PATH}
```
Comment (tonysy) [inline docs/en/user_guides/visualization.md]:
```suggestion
python tools/analysis_tools/visualize_reconstruction.py ${CONFIG_FILE} \
    --checkpoint ${CKPT_PATH} \
    --img-path ${IMAGE_PATH} \
    --out-file ${OUTPUT_PATH}
```

File: docs/en/user_guides/visualization.md [hunk 5]
```diff
@@ -86,3 +89,103 @@ E.g.
...
+- `CONFIG_FILE`: config file for the pre-trained model.
+- `CKPT_PATH`: the path of model's checkpoint.
+- `IMAGE_PATH`: the input image path.
+- `OUTPUT_PATH`: the output image path, including 4 sub-images.
+- `[optional arguments]`: for optional arguments, you can refer to [visualize_reconstruction.py](https://github.com/open-mmlab/mmselfsup/blob/dev-1.x/tools/analysis_tools/visualize_reconstruction.py)
+
+An example:
+
+```shell
+python tools/analysis_tools/visualize_reconstruction.py configs/selfsup/mae/mae_vit-huge-p16_8xb512-amp-coslr-1600e_in1k.py --checkpoint https://download.openmmlab.com/mmselfsup/1.x/mae/mae_vit-huge-p16_8xb512-fp16-coslr-1600e_in1k/mae_vit-huge-p16_8xb512-fp16-coslr-1600e_in1k_20220916-ff848775.pth --img-path data/imagenet/val/ILSVRC2012_val_00000003.JPEG --out-file test_mae.jpg --norm-pix
```
Comment (tonysy) [inline docs/en/user_guides/visualization.md]:
```suggestion
python tools/analysis_tools/visualize_reconstruction.py configs/selfsup/mae/mae_vit-huge-p16_8xb512-amp-coslr-1600e_in1k.py \
    --checkpoint https://download.openmmlab.com/mmselfsup/1.x/mae/mae_vit-huge-p16_8xb512-fp16-coslr-1600e_in1k/mae_vit-huge-p16_8xb512-fp16-coslr-1600e_in1k_20220916-ff848775.pth \
    --img-path data/imagenet/val/ILSVRC2012_val_00000003.JPEG \
    --out-file test_mae.jpg \
    --norm-pix
```

File: docs/en/user_guides/visualization.md [hunk 6]
```diff
@@ -86,3 +89,103 @@ E.g.
...
+
+
+# As for SimMIM, it generates the mask in data pipeline, thus we use '--use-vis-pipeline' to apply 'vis_pipeline' defined in config instead of the pipeline defined in script.
+python tools/analysis_tools/visualize_reconstruction.py configs/selfsup/simmim/simmim_swin-large_16xb128-amp-coslr-800e_in1k-192.py --checkpoint https://download.openmmlab.com/mmselfsup/1.x/simmim/simmim_swin-large_16xb128-amp-coslr-800e_in1k-192/simmim_swin-large_16xb128-amp-coslr-800e_in1k-192_20220916-4ad216d3.pth --img-path data/imagenet/val/ILSVRC2012_val_00000003.JPEG --out-file test_simmim.jpg --use-vis-pipeline
```
Comment (tonysy) [inline docs/en/user_guides/visualization.md]:
```suggestion
python tools/analysis_tools/visualize_reconstruction.py configs/selfsup/simmim/simmim_swin-large_16xb128-amp-coslr-800e_in1k-192.py \
    --checkpoint https://download.openmmlab.com/mmselfsup/1.x/simmim/simmim_swin-large_16xb128-amp-coslr-800e_in1k-192/simmim_swin-large_16xb128-amp-coslr-800e_in1k-192_20220916-4ad216d3.pth \
    --img-path data/imagenet/val/ILSVRC2012_val_00000003.JPEG \
    --out-file test_simmim.jpg \
    --use-vis-pipeline
```

File: docs/en/user_guides/visualization.md (hunk 3 above)
Comment (fangyixiao18) [reply docs/en/user_guides/visualization.md:134]: 这个我这个月重构 tsne 的时候一起改吧，这边也只是把文档位置挪了一下

**Review by tonysy — APPROVED** [review]
LGTM

**Pull request merged**